primitives/src/vid/namespace.rs

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2023 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! Trait for namespace functionality in Verifiable Information Retrieval (VID).
+
+use super::{VidResult, VidScheme};
+
+/// Namespace functionality for [`VidScheme`].
+pub trait Namespacer: VidScheme {
+    /// data proof
+    type DataProof;
+
+    /// Compute a proof for `payload` for data index range `start..start+len-1`.
+    ///
+    /// See TODO in `namespace_verify` on `payload`.
+    fn data_proof(
+        &self,
+        payload: &Self::Payload,
+        start: usize,
+        len: usize,
+    ) -> VidResult<Self::DataProof>;
+
+    /// Verify a proof for `payload` for data index range `start..start+len-1`.
+    ///
+    /// See TODO in `namespace_verify` on `payload`.
+    fn data_verify(
+        &self,
+        payload: &Self::Payload,
+        start: usize,
+        len: usize,
+        commit: &Self::Commit,
+        common: &Self::Common,
+        proof: &Self::DataProof,
+    ) -> VidResult<Result<(), ()>>;
+
+    /// Verify the `payload` namespace indexed by `namespace_index` against
+    /// `commit`, `common`.
+    ///
+    /// TODO: We prefer not to include the whole `payload`. But the namespace
+    /// proof needs a few payload bytes from outside the namespace. In the
+    /// future `payload` should be replaced by a payload subset that includes
+    /// only the bytes needed to verify a namespace.
+    fn namespace_verify(
```

Comment:
I thought we discussed not having any namespace-specific interface in Jellyfish, but having the sequencer itself use the namespace table to convert from namespace index to byte range, and then just use the above APIs to read bytes?

Reply:
My takeaway from that discussion is the rule "no namespace-specific stuff in jellyfish, instead just give me a byte range" applies only to  *transactions*.

Things are a bit different for a full namespace proof. I do suspect that `namespace_verify` in jellyfish will never be used except in tests for correctness. So basically this function is a reference implementation against which to test smart contract verification of a namespace. wdyt?

Reply:
As discussed offline, both namespace and tx proofs will have the exact same API. I'll make a generic trait and implement it twice: once for namespace proofs (snark-friendly) and once for tx proofs (requires a pairing). There will be no namespace-specific stuff in jellyfish. Instead the API is just "give me a index range and I'll give you a proof."